### ISSUE TITLE
Do you have plans to make this plugin complete?

### ISSUE THREAD

**Issue by nazar-pc** [modo-lv/LiveScriptBrains#2]
I mean, is this just an experiment, or you'll continue development?
There is https://github.com/racklin/livescript-idea, but it was abandoned.
LiveScript is awesome, but there is no proper support for it in Jetbrains famili of IDEs.

**Comment by modo-lv** [comment]
It's currently on hiatus because I've tried to to write a decent parser (not just a syntax highlighter like livescript-idea) several times, but I just keep getting stuck. The problem is that LiveScript is a very contextual language where the token you are processing can't be determined until you've processed tokens further on, in addition to being white-space sensitive. I'm very inexperienced with language parsing and WebStorm (IntelliJ) development (this is my first project), so after like the fourth or fifth dead end I kinda gave up, at least for the time being.

**Comment by modo-lv** [comment]
Though I'm not opposed to resuming it sometime, LiveScript is still the best JS version for me (never write projects in anything else if I can help it). Maybe start with something incomplete, like only recognizing simple tokens (variables, operators, keywords), and then try to upgrade step by step.

**Comment by nazar-pc** [comment]
Yeah, I've spent quite some time just to read through all features it has and automatically realized that I'll not use even half of that on daily basis, have no itdea how to write parser for that.
Maybe, some implementations for other IDEs (probably, in other languages) exist and might be used as reference point?

**Comment by modo-lv** [comment]
I haven't found any full-fledged IDE with proper LiveScript support — plugin or otherwise. Most of what I have seen appears to be no more than syntax highlighters. But if I'm going to spend time writing a plugin, I want it to be a bit more. WebStorm has many incredibly useful features, such as, for example, smart renaming (rename a variable / method and automatically & intelligently find and rename it everywhere that it is used) that I'm constantly missing even if the syntax highlighting works.

**Comment by nazar-pc** [comment]
To be honest - what I expect to be good enough is proper syntax highlighting (inheriting from theme defaults) and reformatting support.
Refactoring is great, but for me that 2 things are bare minimum to make it usable. Currently available plugin doesn't even have proper syntax highlighting:(

**Comment by modo-lv** [comment]
Syntax highlighting is easy in that you just define regular expressions for tokens + states for context and let the Lexer take it from there. The IDE doesn't have to "understand" what is written, just switch states & match text for highlighting. Anything more than that (including reformatting) has to go through WebStorm's parser classes, which means, well, parsing. And since their Grammar-Kit format is unusable for something so context- and whitespace-driven as LiveScript (at least as far as I can see), you have to write your own parser. That's what I've attempted to do, but got stuck at some point every time. But thanks for your comments, knowing there is someone else out there looking for the same thing might help me find time to return to this.

**Comment by nazar-pc** [comment]
There are more of us here: https://youtrack.jetbrains.com/issue/WEB-6499 :)

**Comment by modo-lv** [comment]
I have pushed commit 01b773047361355083f4988ccdd3ba0ce1d04072 which contains a (hopefully) working .jar plugin. I dropped the parser for now and focused on gettting at least a little bit of syntax highlighting working. Please try it out if you're interested and let me know what you think.

**Comment by nazar-pc** [comment]
No syntax highlighting at all, no color settings for LiveScript.
Also there is no need to make `*.jar` file, since it is possible to install zip file, and thus zipball of rpeository is perfectly fine for IDE, at least for testing purposes.

**Comment by modo-lv** [comment]
Yes, I see. It works fine in IntelliJ IDEA, but not in WebStorm. I'll see what I can do.

**Comment by nazar-pc** [comment]
I'm using PhpStorm

**Comment by modo-lv** [comment]
I don't have PhpStorm myself, but I added the missing compatibility setting to the META-INF config and it should now work on all IntelliJ Platform based IDEs (if I understand the docs correctly). Commit 5cb9ea0e66d35654a579a28a17f22f109a4ffff5, tested on WebStorm.

**Comment by nazar-pc** [comment]
Very basic, but works now.
Few issues:
```livescript
cs.ui.simple_modal("""
	<h2>#title</h2>
	<cs-system-admin-permissions-for-item label="#{e.model.item.index}" group="Block"/>
""")
```
Comment on third string is not a comment actually.
Also in `.on(..)` method name is highlighted as predefined constant, while it should not be so
Overall this is roughly the same what I was able to achieve with custom language addition.

Hopefully, this plugin will evolve further, thanks for sharing.

**Comment by nazar-pc** [comment]
I'd suggest take snippet from CoffeeScript in color settings window, convert into LiveScript and achieve its proper highlighting, that would be quite good milestone that will allow eyes to relax a bit instead of reading code where everything have the same color.

**Comment by nazar-pc** [comment]
I found some kind of solution to bootstrap development MUCH further!

You can try to call `lsc` (CLI compiler) under the hood and use results in Java code.

Having LiveScript compiler installed seems to be reasonable dependency, also `lsc` might be already availabe in order to use in File Watchers or compile it manually.

This will radically decreate complexity for you, moreover, you'll get AST, lexer and other updates for free since all of those are supported by LiveScript developers. I think this is reasonable dependency to make this plugin alive and do that faster.

**Comment by modo-lv** [comment]
Thanks, that's a great idea. I hadn't noticed that `lsc` can also output lexer tokens. You're right, that would help quite a bit, if I can figure out how to include it in the plugin. :)

**Comment by nazar-pc** [comment]
I think you don't need to include it into plugin, just let users to specify path to executable. It will be different fo Linux/Windows/Mac OS X

**Comment by modo-lv** [comment]
I had a bit of time so I looked in more detail at the `lsc` output and I'm seeing some problems with this approach:
1) How to "connect" character position in the editor with the output of `lsc`? LiveScript compiler parses an entire file and outputs a string of tokens, while the IDE goes through the text character-by-character when building tokens and then token-by-token when parsing it (presumably storing each token's position in the text). `lsc` token output has no character position information.
2) Performance concerns. The file is re-parsed constantly, often several times per second, even when there are no changes being made. If an external, node.js-based compiler is called that often, it sounds like a it would be a major resource hog and slowdown.

**Comment by nazar-pc** [comment]
Performance first: usually everything is updated when user starts typing, namely after some delay (for instance, 300ms) after last character typed if there was no character typed after it; this means that script will not be executed too often (3 times per second max in this case); also I wrote small script and saw that even 10 times per seconds it doesn't make load more than 9% of single core which is quite affordable.

Now about output: it is a bit more complex, but generally you can drop it into characters.
For instance, piece of code from docs:
```coffeescript
table1 =
  * id: 1
    name: 'george'
  * id: 2
    name: 'mike'
  * id: 3
    name: 'donald'

map (.age), table2 |> fold1 (+)
```
Lex output is:
```
NEWLINE:\n ID:table1 ASSIGN:=
INDENT:2 LITERAL:void ASSIGN:= INDENT:2 ID : STRNUM:1
NEWLINE:\n ID:name : STRNUM:'george'
DEDENT:2 NEWLINE:\n LITERAL:void ASSIGN:= INDENT:2 ID : STRNUM:2
NEWLINE:\n ID:name : STRNUM:'mike'
DEDENT:2 NEWLINE:\n LITERAL:void ASSIGN:= INDENT:2 ID : STRNUM:3
NEWLINE:\n ID:name : STRNUM:'donald'

DEDENT:2 DEDENT:2 NEWLINE:\n ID:map ( DOT:. ID:age ) , ID:table2 PIPE:|> ID:fold1 ( BIOP:+ ) NEWLINE:\n
```
So, at the first glance every line of LiveScript corresponds to exactly on line of Lex output, which is awesome.
Seconds thing is that it shows everything meaningful: if there is indentation - you'll see `INDENT:2`, if there is assign symbol - you'll see `ASSIGN:=` (for `:=` it will be `ASSIGN::=`).
Basically you have information about all important tokens, but no information about number of meaningless spaces. However, since you know next meaningful statement in the string you can find it and map entire string of arbitrary statements to whatever data structure you need.

**Comment by modo-lv** [comment]
Doesn't seem that simple. Look at your own example — how does
````
  * id: 1
````
translate into
````
INDENT:2 LITERAL:void ASSIGN:= INDENT:2 ID : STRNUM:1
````
(or back)?

What about code pieces like this?
````
multiline = 'string can be multiline
            and go on and on
            beginning whitespace is
            ignored'
````
Lexer output is
````
NEWLINE:\n ID:multiline ASSIGN:= STRNUM:'string can be multilineand go on and onbeginning whitespace isignored'
````
which means both newlines and spaces are lost, and in this case spaces are not meaningless, they are part of the string.

Same wtih multi-line regex:
````
//
| [!=]==?             # equality
| @@                  # constructor
| <\[(?:[\s\S]*?\]>)? # words
//g
````
which is Lexed as `LITERAL:/|[!=]==?|@@|<\[(?:[\s\S]*?\]>)?/g`

Single-line comments are ignored completely in all cases.

Then there's replacement: `(a and b)` is lexed as `( ID:a LOGIC:&& ID:b )`, showing something that isn't in the text, and of a different length.

I only grabbed a few random examples, but so far it appears enough workarounds and extra functionality would be needed (including rewriting the lexer classes, implementing an `lsc` call and a parser for it's output, and adding a GUI so that users can configure the path) that in the end it would take just as much if not more time to create as a custom lexer, which is already partially done and handling most cases pretty OK. Trying to switch to LiveScript compiler would be less "bootstrapping" and more "throwing out everything already done, doing twice as much and rewriting existing IDE classes for the same functionality and probably less efficient". I might try it out as an experiment (and you're welcome to provide pull requests if you are interested in trying it out), but the first priority for me is having a working plugin, so I'd rather focus what little free time I have on finishing up the current lexer and moving on to tackling the parser next.

**Comment by nazar-pc** [comment]
For me lexer actually returns this:
```
NEWLINE:\n ID:multiline ASSIGN:= STRNUM:'string can be multilineand go on and onbeginning whitespace isignored'


NEWLINE:\n
```
Same story here:
```
NEWLINE:\n



LITERAL:/|[!=]==?|@@|<\[(?:[\s\S]*?\]>)?/g NEWLINE:\n
```
New lines are not actually lost, but output is definitely weird.
**Which seems to be a bug in it. Would you mind to report it upstream and provide link to issue here?**

For now number of lines on input and output is still correct, but you'll need some hacks to find correspondence of multiline text and lexer output.

`and` statement is not a huge problem as well, just needs to be known to your parser.

I my opinion if you write own lexer for such a complex language you'll never be too close to what is officially provided, while having some workarounds for official lexer (like multiline hacks + synonyms for boolean logic + some other) to normalize its output makes more sense for me.

I do not know Java, so except reading I'm not able to help that much in coding itself here.

**Comment by mnn** [comment]
> knowing there is someone else out there looking for the same thing might help me find time to return to this.

I think **there is a lot of others** like me, **waiting for somewhat usable IDE plugin** before committing to a new language. I'd love to switch/start some smaller projects in LiveScript, but without proper (any) IDE support it's discouraging... I either end up with Coffee or ES.

**Comment by nazar-pc** [comment]
Anyway, I've got some very basic syntax highlighting with custom language feature of IDE and already write all new code in LiveScript.
Having one of the most popular IDE currently would definitely bring more users to language itself (I hope it will be even more popular than CoffeeScript, since CoffeeScript have no benefits comparing to LiveScript).
